### ISSUE TITLE
create a pull-request for verificatum

### ISSUE THREAD

**Issue by monperrus** [SpoonLabs/sorald#247]
https://github.com/verificatum/verificatum-vmn

which is the demonstrator in trustfull.

**Comment by slarse** [comment]
This largely depends on #256 being fixed first, as verificatum uses spaces for indentation.

**Comment by slarse** [comment]
This should be possible for next week's milestone

**Comment by slarse** [comment]
Add overview of all violations

**Comment by slarse** [comment]
`verificatum-vmn` is really not very vulnurable, here are all of the violations that our current checks can see:

```
DeadStoreCheck=8
ConditionalUnreachableCodeCheck=6
NullDereferenceCheck=6
InterruptedExceptionCheck=3
CastArithmeticOperandCheck=1
UnclosedResourcesCheck=1
IdenticalOperandOnBinaryExpressionCheck=1
```

Several of these are false positives. For example, the dead stores seem deliberate, as it's setting `null` in variables with sensitive information after they are no longer needed. It's overall just a very carefully written codebase with few flaws, not your average open source project, I'd say.

Henre are the detailed mining stats: [verificatum_vmn_mining_stats.zip](https://github.com/SpoonLabs/sorald/files/5701272/verificatum_vmn_mining_stats.zip)

`verificatum-vcr` has some more stuff:

```
EqualsOverridenWithHashCodeCheck=22
InterruptedExceptionCheck=21
DeadStoreCheck=14
AssertOnBooleanVariableCheck=12
UnclosedResourcesCheck=7
ConditionalUnreachableCodeCheck=6
SymmetricEqualsCheck=5
CastArithmeticOperandCheck=3
MutableMembersUsageCheck=3
ParameterReassignedToCheck=2
NullDereferenceCheck=2
RawByteBitwiseOperationsCheck=1
FloatEqualityCheck=1
NoWayOutLoopCheck=1
XxeProcessingCheck=1
```

Unfortunately, the XxeProcessingCheck seems to be a false positive, as there is an explicitly set external schema, so it does not seem like an oversight on part of the developer (as is usually the case with XxeProcessingCheck).

Here are the detailed mining statistics: [verificatum_vcr_mining_stats.zip](https://github.com/SpoonLabs/sorald/files/5701324/verificatum_vcr_mining_stats.zip)


But in summary, `verificatum-vcr` seems much more amenable for repair than `verificatum-vmn`, so I will proceed with investigating the former.

**Comment by slarse** [comment]
I believe that I have found a true positive.

```java
    public static int optimalWidth(final int bitLength) {

        // This computes the theoretical optimum.
        int width = 1;
        double cost = 1.5 * bitLength;
        double oldCost;
        do {

            oldCost = cost;

            width++;
            final int widthExp = 1 << width;
            cost =
                ((double) (widthExp + (2 - 1 / widthExp) * bitLength)) / width; // noncompliant

        } while (cost < oldCost);

        return Math.max(1, width - 1);
    }
```

```java
            cost =
                ((double) (widthExp + (2 - 1 / widthExp) * bitLength)) / width;
```

is noncompliant with rule `2184: Math operands should be cast before assignment`, because the subexpression `(2 - 1 / widthExp)` has division `1 / widthExp`, which is always 0 as both numerator and denominator are integers. We can patch this, but the output is a bit off:

```diff
             cost =
-                ((double) (widthExp + (2 - 1 / widthExp) * bitLength)) / width;
+                ( (double) (widthExp + ((2 - ((double) 1 / widthExp)) * bitLength))) / width;
```

There is an unwanted space at the beginning of the expression, not sure why that's happening. Not a security fix per se, but it's at least a non-trivial patch.

**Comment by slarse** [comment]
The unwanted space is a bug in the sniper printer, it fails to match a parenthesis to the cast expression, instead matching it to the parenthesis following the cast, thereby printing an unwanted space character.

**Comment by slarse** [comment]
I've poked about in the sniper printer, and fixing this issue with the unwanted space is probably infeasibly difficult. The reason for this is that the sniper printer attempts to match tokens that the default printer wants to print to tokens in the source code. The default printer is extremely liberal with printing parentheses, and accurately matching this abundance of parentheses tokens to actual source tokens is virtually impossible.

In trying to deal with this, I think we can add handling for corner cases ad infinitum. The best way to solve this would probably be to decrease the amount of redundant parentheses that the default printer outputs, especially in binary operator expressions.

**Comment by monperrus** [comment]
> Not a security fix per se, but it's at least a non-trivial patch.

Yes, let's try and see what Douglas thinks of the warning and its fix.

**Comment by slarse** [comment]
After discussion with Douglas, I will move this to next year.

**Comment by slarse** [comment]
First PR is merged: `https://github.com/verificatum/verificatum-vmn/pull/11`

In the future, we are requested to submit one PR _per rule_. In other words, if there are multiple violations of a rule, they should all be fixed in the same PR.

**Comment by monperrus** [comment]
Nice, that's a new passed milestone for us! Congrats.

**Comment by slarse** [comment]
As we've made the first PR for this project, I'm closing this issue. We all know creating PRs is a standing mission for Sorald, so it doesn't have to clutter up the issue tracker :)

I expect to return to making PRs for Verificatum in about 2 weeks time.

**Comment by monperrus** [comment]
We have 36k LOC in https://github.com/verificatum/verificatum-vcr, can Sorald fix something in verificatum-vcr also?

**Comment by slarse** [comment]
> We have 36k LOC in verificatum/verificatum-vcr, can Sorald fix something in verificatum-vcr also?

Yes! I've been meaning to get around to it, but I've simply prioritized other things. Now that you've brought it up again, I'll get to it after I've completed a first draft of the GitHub Action.

Thanks for the reminder :)

**Comment by monperrus** [comment]
great, thanks.

**Comment by slarse** [comment]
See #492
